2067 snapshot based runners seem to only use a single core v13 mapper

benjaperez1983 opened the pull request:
Fixes #2067

# Description

Execute multiple simulations when loaded from snapshot.

## Type of change

Please mark relevant options with an `x` in the brackets.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [ ] Unit tests
- [x] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

rwmcintosh committed:
fix with ben

benjaperez1983 committed:
#2067 snapshot mapper project mapper

benjaperez1983 committed:
#2067 revert .csproj

benjaperez1983 committed:
#2067 format

benjaperez1983 committed:
#2067 format

benjaperez1983 committed:
#2067 fix tests

rwmcintosh commented on src/MoBi.Core/Snapshots/Mappers/ProjectMapper.cs:
```diff
@@ -105,12 +112,13 @@ private async Task<ModelProject> mapToModel(SnapshotProject projectSnapshot, Pro
 
+      List<Task<MoBiSimulation>> tasks = [];
       foreach (var simulationSnapshot in projectSnapshot.Simulations)
       {
          try
          {
-            var simulation = await _simulationMapper.MapToModel(simulationSnapshot, simulationContext);
-            addSimulations(project, simulation);
+            tasks.Add(_simulationMapper.MapToModel(simulationSnapshot, simulationContext));
+
```
I think this should revert to the original code. We don't want to run the construction of simulations in parallel, only the actual simulation run.

rwmcintosh commented on src/MoBi.Core/Snapshots/Mappers/ProjectMapper.cs:
```diff
@@ -119,6 +127,47 @@ private async Task<ModelProject> mapToModel(SnapshotProject projectSnapshot, Pro
+
+      sims.Each(x => addSimulations(project, x));
+
+      if (simulationContext.Run && sims.Any())
+      {
+         var maxParallel = Math.Max(1, _userSettings.MaximumNumberOfCoresToUse);
+         using var gate = new SemaphoreSlim(maxParallel);
+
+         var runTasks = sims.Select(async sim =>
+         {
+            await gate.WaitAsync();
+            try
+            {
+               await _simulationRunner.RunSimulationAsync(sim);
+            }
+            catch (Exception ex)
+            {
+               _logger.AddException(ex);
+            }
+            finally
+            {
+               gate.Release();
+            }
+         });
+
+         await Task.WhenAll(runTasks);
+      }
```
Were you able to investigate `IConcurrentSimulationRunner` from core to see if it can be used? It is very similar to this code.

benjaperez1983 replied:
I did, and also consulted some Microsoft Documentation.

I used SemaphoreSlim + Task.WhenAll instead of the Paralel.Foreach because our workload is async (RunSimulationAsync). 
IConcurrentSimulationRunner is built on sync Parallel.ForEach (wrapped in Task.Run), which would force blocking on each async call.

The correct approach would be to use Parallel.ForEachAsync for this, which would cover our case, but unfortunately it is not available on this framework (.NET 6+).

https://learn.microsoft.com/en-us/training/modules/implement-asynchronous-tasks/5-run-asynchronous-tasks-parallel
https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync?view=net-9.0

The issue is here:

return Task.Run(() => Parallel.ForEach(data, createParallelOptions(cancellationToken, numberOfCoresToUse),
   datum =>
   {
      cancellationToken.ThrowIfCancellationRequested();
      action(datum, cancellationToken);   // <-- ISSUE HERE, this is not an ASYNC Method or task
      // For sims we'd need: _simulationRunner.RunSimulationAsync(sim)
      // but Parallel.ForEach can't await. We must either:
      //  1) fire-and-forget (BUG: ForEach finishes before tasks do, exceptions lost), or
      //  2) block with .GetAwaiter().GetResult() --> Blocks the Thread, deadlocks risk.
   }), cancellationToken);
   
So the real issue is the loop body call (action(...)) inside Parallel.ForEach. 
With _simulationRunner.RunSimulationAsync(sim) there’s no way to await without either not waiting or blocking.
SemaphoreSlim + Task.WhenAll avoids both of these issues.

Let me know your Thoughts about this.

benjaperez1983 committed:
#2067 remove paralell simulation creation

rwmcintosh commented on src/MoBi.Core/Snapshots/Mappers/ProjectMapper.cs:
```diff
@@ -105,12 +112,15 @@ private async Task<ModelProject> mapToModel(SnapshotProject projectSnapshot, Pro
          NumberOfSimulationsLoaded = 0
       };
 
+      var sims = new List<MoBiSimulation>();
```
I guess we could get away without this variable. We only use it to tally the simulations, but we are going to run all the simulations in the project after this step so we could just get it from there.

Yuri05 replied:
Agree

rwmcintosh commented on src/MoBi.Core/Snapshots/Mappers/ProjectMapper.cs:
```diff
@@ -119,6 +129,31 @@ private async Task<ModelProject> mapToModel(SnapshotProject projectSnapshot, Pro
 
+      if (simulationContext.Run && sims.Any())
+      {
+         var maxParallel = Math.Max(1, _userSettings.MaximumNumberOfCoresToUse);
+         using var gate = new SemaphoreSlim(maxParallel);
+
+         var runTasks = sims.Select(async sim =>
+         {
+            await gate.WaitAsync();
+            try
+            {
+               await _simulationRunner.RunSimulationAsync(sim);
+            }
+            catch (Exception ex)
+            {
+               _logger.AddException(ex);
+            }
+            finally
+            {
+               gate.Release();
+            }
+         });
+
+         await Task.WhenAll(runTasks);
```
Could you move it to a method to improve the readability there?

rwmcintosh reviewed (approved):
I would just make a couple small changes.

msevestre commented:
>IConcurrentSimulationRunner is built on sync Parallel.ForEach (wrapped in Task.Run), which would force blocking on each async call.

I don't understand what you mean. Parallel.ForEach blocks on each call?

msevestre commented:
>The correct approach would be to use Parallel.ForEachAsync for this, which would cover our case, but unfortunately it is not available on this framework (.NET 6+).

That's an interesting question. We are moving to .NET 8 now. Should we move our components when needed or is the implication of this way too high?

rwmcintosh commented:
> >The correct approach would be to use Parallel.ForEachAsync for this, which would cover our case, but unfortunately it is not available on this framework (.NET 6+).
> 
> 
> 
> That's an interesting question. We are moving to .NET 8 now. Should we move our components when needed or is the implication of this way too high?

At least testing here limits the reach to mobi. @benjaperez1983 can you try it? Do all other assemblies need to be upgraded at the same time?

benjaperez1983 commented:
> IConcurrentSimulationRunner

For what I`ve read about Parallel.ForEach is a **synchronous** API. 
It runs a **sync delegate** on multiple threads and the call doesn’t return until all iterations finish. 
Because the delegate is sync, if I need to call an async method like RunSimulationAsync, I have only two choices:

1) Block the async call so it looks sync:

Parallel.ForEach(sims, opt, sim =>
{
  _simulationRunner.RunSimulationAsync(sim).GetAwaiter().GetResult(); // blocking a worker thread
});


which “forces blocking” (what I meant): each iteration ties up a thread while the async operation is running.

2) Fire-and-forget inside the body (don’t wait), which is incorrect because Parallel.ForEach would finish before the tasks do, and exceptions/cancellation are lost.

Task.Run(() => Parallel.ForEach(...)) doesn’t make the work asynchronous, it just runs our synchronous loop on a thread-pool thread. While it’s running, threads are still blocked.

benjaperez1983 commented:
Converted this to draft, to prevent any merging. Since once I implemented the Parallel.ForEachAsync, I can see other changes need to be made at Core.
So I`d rather discuss them with @rwmcintosh  or all the team before merging.

benjaperez1983 committed:
#2067 parallel foreach

rwmcintosh committed:
new simmodelmanager per simulation run

benjaperez1983 committed:
#2067 framework change

benjaperez1983 commented:
depends on https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/pull/2659

benjaperez1983 committed:
#2607 changes for PR

benjaperez1983 committed:
merge with v13

benjaperez1983 committed:
#2067 remove extensions and refactor

benjaperez1983 committed:
#2067 Remove Enumerable Extension Method

rwmcintosh reviewed (changes requested):
The document submodule change should be reverted

benjaperez1983 committed:
#2067 revert documentation

benjaperez1983 commented:
> The document submodule change should be reverted

Good Catch! already reverted it to V13

rwmcintosh commented on src/MoBi.Presentation/Tasks/ObservedDataTask.cs:
```diff
@@ -354,7 +354,12 @@ private void addPredefinedMoleculesForImporter(MetaDataCategory metaDataCategory
+         return System.Linq.Enumerable.DistinctBy(
+            _buildingBlockRepository.MoleculeBlockCollection.SelectMany(bb => bb),
+            b => b.Name,
+            StringComparer.Ordinal);
```
that should go back to the extension method syntax rather than the static method syntax right? Only change is who is providing the extension?

benjaperez1983 replied:
Yeah I specified the namespace because of the ambiguous call. 
But reverting it.

msevestre reviewed (approved):
loooks good. Just one change with the extgension methods not called properly

benjaperez1983 committed:
#2067 distinct by
